File: tools/davinci/matrix-gen.mjs
// Construct-matrix fixture generator — SKELETON (Davinci P0-12).
//
// Reads davinci-road/plan/taxonomy.toml (DRAFT, pending maintainer
// sign-off) and emits deterministic .vue fixture stubs under
// tests/fixtures/davinci-matrix/. This skeleton generates the
// element_kind x directive plane only; the full cross-product (modifier
// classes, binding sources, block combinations) arrives with the pipeline
// stages that consume the fixtures, which also bring the expected
// outputs. Stubs are source text only — no expected outputs, nothing is
// compiled here.
//
// Modes:
//   (default)        dry run — validate the taxonomy and print the
//                    would-be fixture count; nothing is written
//   --write          write the stubs to the target directory
//   --check          verify the on-disk fixture set byte-matches a fresh
//                    generation (exit 1 listing missing/stale/extra files)
//   --out-dir <dir>  override the target directory (self-tests write to a
//                    temp dir; the committed home is tests/fixtures/davinci-matrix)
//
// Determinism contract (assurance doctrine, "generated fixtures, not
// hand-picked examples"): output depends only on taxonomy.toml. Iteration
// follows taxonomy file order, stub content carries no timestamps,
// absolute paths, or environment data — two runs are byte-identical.
//
// Exit codes: 0 = success / up to date, 1 = --check found drift,
// 2 = usage or taxonomy validation error.

import fs from "node:fs";
import path from "node:path";
import process from "node:process";
import { fileURLToPath } from "node:url";

import { parseTomlLite, TomlLiteError } from "./toml-lite.mjs";

const repoRoot = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..", "..");
const taxonomyPath = path.join(repoRoot, "davinci-road", "plan", "taxonomy.toml");
const defaultOutDir = path.join(repoRoot, "tests", "fixtures", "davinci-matrix");

const ID_RE = /^[a-z][a-z0-9-]*$/;
const DIMENSIONS = [
  "element_kind",
  "directive",
  "modifier_class",
  "binding_source",
  "block_combination",
];

function fail(message) {
  console.error(message);
  process.exit(2);
}

function loadTaxonomy() {
  let parsed;
  try {
    parsed = parseTomlLite(fs.readFileSync(taxonomyPath, "utf8"));
  } catch (error) {
    if (error instanceof TomlLiteError)
      fail(`malformed taxonomy ${taxonomyPath}: ${error.message}`);
    throw error;
  }
  for (const dimension of DIMENSIONS) {
    const entries = parsed[dimension];
    if (!Array.isArray(entries) || entries.length === 0) {
      fail(`taxonomy ${taxonomyPath}: missing [[${dimension}]] entries`);
    }
    const ids = new Set();
    for (const entry of entries) {
      if (typeof entry.id !== "string" || !ID_RE.test(entry.id)) {
        fail(`taxonomy [[${dimension}]]: entry with missing or non-kebab-case id`);
      }
      if (ids.has(entry.id)) fail(`taxonomy [[${dimension}]]: duplicate id ${entry.id}`);
      ids.add(entry.id);
    }
  }
  for (const kind of parsed.element_kind) {
    if (typeof kind.representative !== "string" || kind.representative.length === 0) {
      fail(`taxonomy [[element_kind]] ${kind.id}: representative tag is required`);
    }
  }
  for (const directive of parsed.directive) {
    if (typeof directive.usage !== "string" || directive.usage.length === 0) {
      fail(`taxonomy [[directive]] ${directive.id}: usage attribute text is required`);
    }
  }
  return parsed;
}

// One stub per element_kind x directive pair, keyed by file name, in
// taxonomy order. Content is pure text derived from the two entries.
export function generateStubs(taxonomy) {
  const files = new Map();
  for (const kind of taxonomy.element_kind) {
    for (const directive of taxonomy.directive) {
      const name = `${kind.id}--${directive.id}.vue`;
      const tag = kind.representative;
      const lines = [
        "<!--",
        "  generated-by: tools/davinci/matrix-gen.mjs (P0-12 skeleton) — do not edit",
        `  construct: element_kind=${kind.id} directive=${directive.id}`,
        "-->",
        "<template>",
      ];
      if (directive.needs_prior_branch === true) {
        lines.push('  <div v-if="first"></div>');
      }
      lines.push(`  <${tag} ${directive.usage}></${tag}>`);
      lines.push("</template>");
      files.set(name, `${lines.join("\n")}\n`);
    }
  }
  return files;
}

function parseArgs(argv) {
  const args = { write: false, check: false, outDir: null };
  for (let i = 0; i < argv.length; i += 1) {
    const arg = argv[i];
    if (arg === "--write") args.write = true;
    else if (arg === "--check") args.check = true;
    else if (arg === "--out-dir") {
      args.outDir = argv[i + 1];
      i += 1;
      if (args.outDir == null) fail("--out-dir requires a directory argument");
    } else if (arg === "--help" || arg === "-h") args.help = true;
    else fail(`unknown argument ${arg}`);
  }
  if (args.write && args.check) fail("--write and --check are mutually exclusive");
  return args;
}

const USAGE = `Usage: rust-script tools/commands/davinci/matrix-gen.rs [--write | --check] [--out-dir <dir>]

Generates construct-matrix fixture stubs from davinci-road/plan/taxonomy.toml.
Default is a dry run that prints the would-be fixture count.`;

function describe(taxonomy) {
  const counts = DIMENSIONS.map((dimension) => `${dimension}=${taxonomy[dimension].length}`);
  console.log(`taxonomy: ${counts.join(" ")} (${taxonomy.taxonomy?.status ?? "unknown"})`);
}

function relative(target) {
  const rel = path.relative(repoRoot, target);
  return rel.startsWith("..") ? target : rel.split(path.sep).join("/");
}

function main() {
  const args = parseArgs(process.argv.slice(2));
  if (args.help) {
    console.log(USAGE);
    return 0;
  }
  const taxonomy = loadTaxonomy();
  const stubs = generateStubs(taxonomy);
  const outDir = path.resolve(args.outDir ?? defaultOutDir);
  describe(taxonomy);
  const plane = `${taxonomy.element_kind.length} element kinds x ${taxonomy.directive.length} directives`;

  if (args.check) {
    const missing = [];
    const stale = [];
    for (const [name, content] of stubs) {
      const target = path.join(outDir, name);
      if (!fs.existsSync(target)) missing.push(name);
      else if (fs.readFileSync(target, "utf8") !== content) stale.push(name);
    }
    const extra = (fs.existsSync(outDir) ? fs.readdirSync(outDir) : [])
      .filter((name) => name.endsWith(".vue") && !stubs.has(name))
      .sort();
    if (missing.length + stale.length + extra.length > 0) {
      for (const name of missing) console.log(`missing: ${name}`);
      for (const name of stale) console.log(`stale: ${name}`);
      for (const name of extra) console.log(`extra: ${name}`);
      console.log(`matrix-gen: ${relative(outDir)} is out of date — rerun with --write and commit`);
      return 1;
    }
    console.log(`matrix-gen: ${stubs.size} fixture stubs up to date in ${relative(outDir)}`);
    return 0;
  }

  if (args.write) {
    fs.mkdirSync(outDir, { recursive: true });
    for (const [name, content] of stubs) {
      fs.writeFileSync(path.join(outDir, name), content);
    }
    const extra = fs
      .readdirSync(outDir)
      .filter((name) => name.endsWith(".vue") && !stubs.has(name))
      .sort();
    console.log(`matrix-gen: wrote ${stubs.size} fixture stubs (${plane}) to ${relative(outDir)}`);
    for (const name of extra) {
      console.warn(`warning: ${name} is not part of the generated set (would fail --check)`);
    }
    return 0;
  }

  console.log(`matrix-gen (skeleton): ${plane} -> ${stubs.size} fixture stubs`);
  console.log(
    `dry run — nothing written; target ${relative(outDir)} (--write to emit, --check to verify)`,
  );
  return 0;
}

const invokedDirectly =
  process.argv[1] != null && fileURLToPath(import.meta.url) === path.resolve(process.argv[1]);
if (invokedDirectly) {
  process.exitCode = main();
}

File: tools/commands/davinci/matrix-gen.rs
#!/usr/bin/env rust-script
//! ```cargo
//! [dependencies]
//! regex = "1"
//! serde = { version = "1", features = ["derive"] }
//! serde_json = "1"
//! toml = "0.9"
//!
//! [package]
//! edition = "2024"
//! ```

#[path = "../../rust/common.rs"]
mod common;

use regex::Regex;
use std::{
    collections::{BTreeMap, BTreeSet},
    env, fs,
    path::{Path, PathBuf},
    process::ExitCode,
};
use toml::Value;

const DIMENSIONS: &[&str] = &[
    "element_kind",
    "directive",
    "modifier_class",
    "binding_source",
    "block_combination",
];
const USAGE: &str = "Usage: rust-script tools/commands/davinci/matrix-gen.rs [--write | --check] [--out-dir <dir>]\n\nGenerates construct-matrix fixture stubs from davinci-road/plan/taxonomy.toml.\nDefault is a dry run that prints the would-be fixture count.";

#[derive(Clone, Debug)]
struct Args {
    write: bool,
    check: bool,
    help: bool,
    out_dir: Option<PathBuf>,
}

#[derive(Clone, Debug)]
struct ElementKind {
    id: String,
    representative: String,
}

#[derive(Clone, Debug)]
struct Directive {
    id: String,
    usage: String,
    needs_prior_branch: bool,
}

#[derive(Clone, Debug)]
struct Taxonomy {
    status: String,
    element_kind: Vec<ElementKind>,
    directive: Vec<Directive>,
    dimensions: BTreeMap<String, usize>,
}

fn main() -> ExitCode {
    match run() {
        Ok(code) => ExitCode::from(code),
        Err((code, error)) => {
            eprintln!("{error}");
            ExitCode::from(code)
        }
    }
}

fn run() -> Result<u8, (u8, String)> {
    let args = parse_args(env::args().skip(1).collect())?;
    if args.help {
        println!("{USAGE}");
        return Ok(0);
    }
    let root = repo_root().map_err(|error| (2, error))?;
    let taxonomy = load_taxonomy(&root).map_err(|error| (2, error))?;
    let stubs = generate_stubs(&taxonomy);
    let out_dir = args
        .out_dir
        .unwrap_or_else(|| root.join("tests/fixtures/davinci-matrix"));
    describe(&taxonomy);
    let plane = format!(
        "{} element kinds x {} directives",
        taxonomy.element_kind.len(),
        taxonomy.directive.len()
    );
    if args.check {
        let mut missing = Vec::new();
        let mut stale = Vec::new();
        for (name, content) in &stubs {
            let target = out_dir.join(name);
            if !target.exists() {
                missing.push(name.clone());
            } else if common::read_text(&target).map_err(|error| (2, error))? != *content {
                stale.push(name.clone());
            }
        }
        let extra = if out_dir.exists() {
            fs::read_dir(&out_dir)
                .map_err(|error| (2, format!("cannot read {}: {error}", out_dir.display())))?
                .filter_map(Result::ok)
                .filter_map(|entry| {
                    let name = entry.file_name().to_string_lossy().into_owned();
                    (name.ends_with(".vue") && !stubs.contains_key(&name)).then_some(name)
                })
                .collect::<BTreeSet<_>>()
        } else {
            BTreeSet::new()
        };
        if missing.len() + stale.len() + extra.len() > 0 {
            for name in missing {
                println!("missing: {name}");
            }
            for name in stale {
                println!("stale: {name}");
            }
            for name in extra {
                println!("extra: {name}");
            }
            println!(
                "matrix-gen: {} is out of date - rerun with --write and commit",
                relative(&root, &out_dir)
            );
            return Ok(1);
        }
        println!(
            "matrix-gen: {} fixture stubs up to date in {}",
            stubs.len(),
            relative(&root, &out_dir)
        );
        return Ok(0);
    }
    if args.write {
        fs::create_dir_all(&out_dir)
            .map_err(|error| (2, format!("cannot create {}: {error}", out_dir.display())))?;
        for (name, content) in &stubs {
            common::write_text(out_dir.join(name), content).map_err(|error| (2, error))?;
        }
        let extra = fs::read_dir(&out_dir)
            .map_err(|error| (2, format!("cannot read {}: {error}", out_dir.display())))?
            .filter_map(Result::ok)
            .filter_map(|entry| {
                let name = entry.file_name().to_string_lossy().into_owned();
                (name.ends_with(".vue") && !stubs.contains_key(&name)).then_some(name)
            })
            .collect::<BTreeSet<_>>();
        println!(
            "matrix-gen: wrote {} fixture stubs ({plane}) to {}",
            stubs.len(),
            relative(&root, &out_dir)
        );
        for name in extra {
            eprintln!("warning: {name} is not part of the generated set (would fail --check)");
        }
        return Ok(0);
    }
    println!(
        "matrix-gen (skeleton): {plane} -> {} fixture stubs",
        stubs.len()
    );
    println!(
        "dry run - nothing written; target {} (--write to emit, --check to verify)",
        relative(&root, &out_dir)
    );
    Ok(0)
}

fn parse_args(argv: Vec<String>) -> Result<Args, (u8, String)> {
    let mut args = Args {
        write: false,
        check: false,
        help: false,
        out_dir: None,
    };
    let mut index = 0;
    while index < argv.len() {
        match argv[index].as_str() {
            "--write" => args.write = true,
            "--check" => args.check = true,
            "--out-dir" => {
                index += 1;
                let value = argv
                    .get(index)
                    .ok_or_else(|| (2, "--out-dir requires a directory argument".to_string()))?;
                args.out_dir = Some(PathBuf::from(value));
            }
            "--help" | "-h" => args.help = true,
            arg => return Err((2, format!("unknown argument {arg}"))),
        }
        index += 1;
    }
    if args.write && args.check {
        return Err((2, "--write and --check are mutually exclusive".to_string()));
    }
    Ok(args)
}

fn load_taxonomy(root: &Path) -> Result<Taxonomy, String> {
    let taxonomy_path = root.join("davinci-road/plan/taxonomy.toml");
    let value: Value = toml::from_str(&common::read_text(&taxonomy_path)?)
        .map_err(|error| format!("malformed taxonomy {}: {error}", taxonomy_path.display()))?;
    let id_re = Regex::new(r"^[a-z][a-z0-9-]*$").unwrap();
    let mut dimensions = BTreeMap::new();
    for dimension in DIMENSIONS {
        let entries = value
            .get(*dimension)
            .and_then(Value::as_array)
            .ok_or_else(|| {
                format!(
                    "taxonomy {}: missing [[{dimension}]] entries",
                    taxonomy_path.display()
                )
            })?;
        if entries.is_empty() {
            return Err(format!(
                "taxonomy {}: missing [[{dimension}]] entries",
                taxonomy_path.display()
            ));
        }
        let mut ids = BTreeSet::new();
        for entry in entries {
            let id = entry
                .get("id")
                .and_then(Value::as_str)
                .filter(|id| id_re.is_match(id))
                .ok_or_else(|| {
                    format!("taxonomy [[{dimension}]]: entry with missing or non-kebab-case id")
                })?;
            if !ids.insert(id.to_string()) {
                return Err(format!("taxonomy [[{dimension}]]: duplicate id {id}"));
            }
        }
        dimensions.insert((*dimension).to_string(), entries.len());
    }
    let element_kind = value
        .get("element_kind")
        .and_then(Value::as_array)
        .unwrap()
        .iter()
        .map(|entry| {
            let id = entry.get("id").and_then(Value::as_str).unwrap().to_string();
            let representative = entry
                .get("representative")
                .and_then(Value::as_str)
                .filter(|value| !value.is_empty())
                .ok_or_else(|| {
                    format!("taxonomy [[element_kind]] {id}: representative tag is required")
                })?
                .to_string();
            Ok(ElementKind { id, representative })
        })
        .collect::<Result<Vec<_>, String>>()?;
    let directive = value
        .get("directive")
        .and_then(Value::as_array)
        .unwrap()
        .iter()
        .map(|entry| {
            let id = entry.get("id").and_then(Value::as_str).unwrap().to_string();
            let usage = entry
                .get("usage")
                .and_then(Value::as_str)
                .filter(|value| !value.is_empty())
                .ok_or_else(|| {
                    format!("taxonomy [[directive]] {id}: usage attribute text is required")
                })?
                .to_string();
            Ok(Directive {
                id,
                usage,
                needs_prior_branch: entry
                    .get("needs_prior_branch")
                    .and_then(Value::as_bool)
                    .unwrap_or(false),
            })
        })
        .collect::<Result<Vec<_>, String>>()?;
    let status = value
        .get("taxonomy")
        .and_then(Value::as_table)
        .and_then(|table| table.get("status"))
        .and_then(Value::as_str)
        .unwrap_or("unknown")
        .to_string();
    Ok(Taxonomy {
        status,
        element_kind,
        directive,
        dimensions,
    })
}

fn generate_stubs(taxonomy: &Taxonomy) -> BTreeMap<String, String> {
    let mut files = BTreeMap::new();
    for kind in &taxonomy.element_kind {
        for directive in &taxonomy.directive {
            let name = format!("{}--{}.vue", kind.id, directive.id);
            let mut lines = vec![
                "<!--".to_string(),
                "  generated-by: tools/commands/davinci/matrix-gen.rs (P0-12 skeleton) \u{2014} do not edit"
                    .to_string(),
                format!("  construct: element_kind={} directive={}", kind.id, directive.id),
                "-->".to_string(),
                "<template>".to_string(),
            ];
            if directive.needs_prior_branch {
                lines.push("  <div v-if=\"first\"></div>".to_string());
            }
            lines.push(format!(
                "  <{} {}></{}>",
                kind.representative, directive.usage, kind.representative
            ));
            lines.push("</template>".to_string());
            files.insert(name, format!("{}\n", lines.join("\n")));
        }
    }
    files
}

fn describe(taxonomy: &Taxonomy) {
    let counts = DIMENSIONS
        .iter()
        .map(|dimension| format!("{dimension}={}", taxonomy.dimensions[*dimension]))
        .collect::<Vec<_>>()
        .join(" ");
    println!("taxonomy: {counts} ({})", taxonomy.status);
}

fn relative(root: &Path, target: &Path) -> String {
    target
        .strip_prefix(root)
        .ok()
        .map(common::normalize_path)
        .unwrap_or_else(|| target.display().to_string())
}

fn repo_root() -> Result<PathBuf, String> {
    common::repo_root().or_else(|_| {
        Path::new(file!())
            .ancestors()
            .find(|candidate| {
                candidate.join("Cargo.toml").is_file()
                    && candidate.join("pnpm-workspace.yaml").is_file()
            })
            .map(Path::to_path_buf)
            .ok_or_else(|| "cannot resolve Vize repository root from script path".to_string())
    })
}
